### PR TITLE
feature: add display for current subscription and cancel button

### PR DESCRIPTION
**Pull request by BarryThePenguin**

### CODE REVIEW

File: lib/graphql/models/stripe.ts [hunk 1]
```diff
@@ -23,3 +43,87 @@ export async function createCheckoutSession(
...
+    expand: ['subscription']
+  });
+
+  assertCustomer(customer);
+
+  // NOTE: Assuming the current subscription is the first active subscription
+  return customer.subscriptions.data.find(
+    subscription => subscription.status === 'active'
+  );
+}
```
Comment (BarryThePenguin) [inline lib/graphql/models/stripe.ts]: These are the important changes

File: lib/graphql/models/stripe.ts [hunk 2]
```diff
@@ -23,3 +43,87 @@ export async function createCheckoutSession(
...
+  const customer = await stripe.customers.retrieve(user.stripeCustomerId, {
+    expand: ['subscription']
+  });
+
+  assertCustomer(customer);
+
+  // NOTE: Assuming the current subscription is the first active subscription
+  return customer.subscriptions.data.find(
+    subscription => subscription.status === 'active'
+  );
```
Comment (BarryThePenguin) [inline lib/graphql/models/stripe.ts]: I'm assuming everyone will only have 1 active subscription

File: lib/graphql/models/stripe.ts [hunk 3]
```diff
@@ -23,3 +43,87 @@ export async function createCheckoutSession(
...
+  return {
+    id: subscription.id,
+    plan: subscription.plan.nickname,
+    status: subscription.status,
+    startDate: fromUnixTime(subscription.start_date),
+    daysUntilDue: subscription.days_until_due,
+    cancelAt: fromUnixTime(subscription.cancel_at),
+    canceledAt: fromUnixTime(subscription.canceled_at),
+    currentPeriodEnd: fromUnixTime(subscription.current_period_end)
+  };
```
Comment (BarryThePenguin) [inline lib/graphql/models/stripe.ts]: If we want to show any more subscription information to the user we need to add them here, and to the [schema.ts](https://github.com/NewFrontDoor/pwad.org.au/blob/master/lib/graphql/schema.ts)

File: src/components/account/manage-account-form.tsx [hunk 4]
```diff
@@ -1,37 +1,57 @@
...
+        }
+      });
+    }
+  });
+
+  async function handleCancelSubscription(): Promise<void> {
+    // eslint-disable-next-line no-alert
+    if (window.confirm('Are you sure you want to cancel your subscription')) {
+      await cancelSubscription();
+    }
```
Comment (BarryThePenguin) [inline src/components/account/manage-account-form.tsx]: I'm just showing a confirmation alert when they click the "Cancel Subscription" button

File: src/components/account/manage-account-form.tsx [hunk 5]
```diff
@@ -48,26 +68,112 @@ const AccountPaymentButton: FC<AccountPaymentButton> = ({
...
+          </>
+        )}
+
+        {daysUntilDue !== null && (
+          <>
+            <dt>Days remaining:</dt>
+            <dd>{daysUntilDue}</dd>
+          </>
+        )}
+      </dl>
```
Comment (BarryThePenguin) [inline src/components/account/manage-account-form.tsx]: This can probably be made to look a bit nicer 😅

File: lib/graphql/models/stripe.ts [hunk 6]
```diff
@@ -23,3 +43,86 @@ export async function createCheckoutSession(
...
+): StripeSubscription {
+  return {
+    id: subscription.id,
+    plan: subscription.plan.nickname,
+    status: subscription.status,
+    startDate: fromUnixTime(subscription.start_date),
+    cancelAt: fromUnixTime(subscription.cancel_at),
+    canceledAt: fromUnixTime(subscription.canceled_at),
+    currentPeriodEnd: fromUnixTime(subscription.current_period_end)
+  };
```
Comment (BarryThePenguin) [inline lib/graphql/models/stripe.ts]: If we want to show any more subscription information to the user we need to add them here, and to the [schema.ts](https://github.com/NewFrontDoor/pwad.org.au/blob/master/lib/graphql/schema.ts)

File: lib/graphql/models/stripe.ts [hunk 7]
```diff
@@ -23,3 +43,86 @@ export async function createCheckoutSession(
...
+    expand: ['subscription']
+  });
+
+  assertCustomer(customer);
+
+  // NOTE: Assuming the current subscription is the first active subscription
+  return customer.subscriptions.data.find(
+    subscription => subscription.status === 'active'
+  );
+}
```
Comment (BarryThePenguin) [inline lib/graphql/models/stripe.ts]: These are the important changes

File: src/components/account/manage-account-form.tsx [hunk 8]
```diff
@@ -48,26 +68,103 @@ const AccountPaymentButton: FC<AccountPaymentButton> = ({
...
+          <>
+            <dt>Subscription canceled at:</dt>
+            <dd>{new Date(canceledAt).toLocaleDateString()}</dd>
+          </>
+        )}
+      </dl>
+
+      {active && <CancelSubscriptionButton />}
+    </Box>
+  );
```
Comment (BarryThePenguin) [inline src/components/account/manage-account-form.tsx]: This can probably be made to look a bit nicer 😅

File: lib/graphql/models/stripe.ts [hunk 9]
```diff
@@ -23,3 +43,86 @@ export async function createCheckoutSession(
...
+ * Cancels the current active subscription for the current user
+ * @param  user The current user
+ * @return      The canceled subscription response
+ */
+export async function cancelSubscription(
+  user: User
+): Promise<StripeSubscription> {
+  const subscription = await findCurrentSubscription(user);
+
+  const deletedSubscription = await stripe.subscriptions.del(subscription.id);
```
Comment (readeral) [inline lib/graphql/models/stripe.ts]: This isn't the method we want. Should swap it for:
```
const cancelledSubscription = await stripe.subscriptions.update(subscription.id, {cancelAtPeriodEnd: true});

return subscriptionResponse(cancelledSubscription)
```

Comment (BarryThePenguin) [reply lib/graphql/models/stripe.ts]: Ooo.. nice! Thanks 👍🏻

**Comment by readeral** [comment]
winner. nice work.

**Pull request merged**